java.lang.NullPointerException:    at openfoodfacts.github.scrachx.openfood.fragments.AddProductIngredientsFragment

``` 
Caused by: java.lang.NullPointerException: 
  at openfoodfacts.github.scrachx.openfood.fragments.AddProductIngredientsFragment.a (AddProductIngredientsFragment.java:70)
  at openfoodfacts.github.scrachx.openfood.views.AddProductActivity.c (AddProductActivity.java:2)
  at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.a (AddProductOverviewFragment.java:83)
  at openfoodfacts.github.scrachx.openfood.fragments.AddProductOverviewFragment$3.a (AddProductOverviewFragment.java:2)
  at io.reactivex.internal.operators.single.SingleObserveOn$ObserveOnSingleObserver.run (SingleObserveOn.java:14)
  at io.reactivex.android.schedulers.HandlerScheduler$ScheduledRunnable.run (HandlerScheduler.java:2)
  at android.os.Handler.handleCallback (Handler.java:808)
  at android.os.Handler.dispatchMessage (Handler.java:101)
  at android.os.Looper.loop (Looper.java:166)
  at android.app.ActivityThread.main (ActivityThread.java:7529)
  at java.lang.reflect.Method.invoke (Method.java)
  at com.android.internal.os.Zygote$MethodAndArgsCaller.run (Zygote.java:245)
  at com.android.internal.os.ZygoteInit.main (ZygoteInit.java:921)
```

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AddProductOverviewFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/AddProductOverviewFragment.java
@@ -462,48 +462,28 @@ public class AddProductOverviewFragment extends BaseFragment {
             if (productDetails.get(PARAM_QUANTITY) != null) {
                 quantity.setText(productDetails.get(PARAM_QUANTITY));
             }
-            if (productDetails.get(PARAM_BRAND) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_BRAND).split("\\s*,\\s*"));
-                brand.setText(chipValues);
-            }
-            if (productDetails.get(PARAM_PACKAGING) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_PACKAGING).split("\\s*,\\s*"));
-                packaging.setText(chipValues);
-            }
-            if (productDetails.get(PARAM_CATEGORIES) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_CATEGORIES).split("\\s*,\\s*"));
-                categories.setText(chipValues);
-            }
-            if (productDetails.get(PARAM_LABELS) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_LABELS).split("\\s*,\\s*"));
-                label.setText(chipValues);
-            }
-            if (productDetails.get(PARAM_ORIGIN) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_ORIGIN).split("\\s*,\\s*"));
-                originOfIngredients.setText(chipValues);
-            }
+            addChipsText(productDetails, PARAM_BRAND, brand);
+            addChipsText(productDetails, PARAM_PACKAGING, packaging);
+            addChipsText(productDetails, PARAM_CATEGORIES, categories);
+            addChipsText(productDetails, PARAM_LABELS, label);
+            addChipsText(productDetails, PARAM_ORIGIN, originOfIngredients);
             if (productDetails.get(PARAM_MANUFACTURING_PLACE) != null) {
                 manufacturingPlace.setText(productDetails.get(PARAM_MANUFACTURING_PLACE));
             }
-            if (productDetails.get(PARAM_EMB_CODE) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_EMB_CODE).split("\\s*,\\s*"));
-                embCode.setText(chipValues);
-            }
+            addChipsText(productDetails, PARAM_EMB_CODE, embCode);
             if (productDetails.get(PARAM_LINK) != null) {
                 link.setText(productDetails.get(PARAM_LINK));
             }
-            if (productDetails.get(PARAM_PURCHASE) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_PURCHASE).split("\\s*,\\s*"));
-                countryWherePurchased.setText(chipValues);
-            }
-            if (productDetails.get(PARAM_STORE) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_STORE).split("\\s*,\\s*"));
-                stores.setText(chipValues);
-            }
-            if (productDetails.get(PARAM_COUNTRIES) != null) {
-                List<String> chipValues = Arrays.asList(productDetails.get(PARAM_COUNTRIES).split("\\s*,\\s*"));
-                countriesWhereSold.setText(chipValues);
-            }
+            addChipsText(productDetails, PARAM_PURCHASE, countryWherePurchased);
+            addChipsText(productDetails, PARAM_STORE, stores);
+            addChipsText(productDetails, PARAM_COUNTRIES, countriesWhereSold);
+        }
+    }
+
+    private void addChipsText(HashMap<String, String> productDetails, String paramName, NachoTextView nachoTextView) {
+        if (productDetails.get(paramName) != null) {
+            List<String> chipValues = Arrays.asList(productDetails.get(paramName).split("\\s*,\\s*"));
+            nachoTextView.setText(chipValues);
         }
     }
 
@@ -730,60 +710,61 @@ public class AddProductOverviewFragment extends BaseFragment {
     public void getDetails() {
         chipifyAllUnterminatedTokens();
         if (activity instanceof AddProductActivity) {
+            final AddProductActivity addProductActivity = (AddProductActivity) this.activity;
             if (!code.isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_BARCODE, code);
+                addProductActivity.addToMap(PARAM_BARCODE, code);
             }
             if (!appLanguageCode.isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_INTERFACE_LANGUAGE, appLanguageCode);
+                addProductActivity.addToMap(PARAM_INTERFACE_LANGUAGE, appLanguageCode);
             }
             if (!languageCode.isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_LANGUAGE, languageCode);
+                addProductActivity.addToMap(PARAM_LANGUAGE, languageCode);
             }
             if (!name.getText().toString().isEmpty()) {
                 String lc = (!languageCode.isEmpty()) ? languageCode : "en";
-                ((AddProductActivity) activity).addToMap(PARAM_NAME + "_" + lc, name.getText().toString());
+                addProductActivity.addToMap(PARAM_NAME + "_" + lc, name.getText().toString());
             }
             if (!quantity.getText().toString().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_QUANTITY, quantity.getText().toString());
+                addProductActivity.addToMap(PARAM_QUANTITY, quantity.getText().toString());
             }
             if (!brand.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_BRAND, getValues(brand));
+                addProductActivity.addToMap(PARAM_BRAND, getValues(brand));
             }
             if (!packaging.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_PACKAGING, getValues(packaging));
+                addProductActivity.addToMap(PARAM_PACKAGING, getValues(packaging));
             }
             if (!categories.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_CATEGORIES, getValues(categories));
+                addProductActivity.addToMap(PARAM_CATEGORIES, getValues(categories));
             }
             if (!label.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_LABELS, getValues(label));
+                addProductActivity.addToMap(PARAM_LABELS, getValues(label));
             }
             if (!periodsAfterOpening.getText().toString().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_PERIODS_AFTER_OPENING, periodsAfterOpening.getText().toString());
+                addProductActivity.addToMap(PARAM_PERIODS_AFTER_OPENING, periodsAfterOpening.getText().toString());
             }
-            if (mImageUrl != null) {
-                ((AddProductActivity) activity).addToMap("imageUrl", mImageUrl);
+            if(mImageUrl!=null){
+                addProductActivity.addToMap("imageUrl", mImageUrl);
             }
             if (!originOfIngredients.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_ORIGIN, getValues(originOfIngredients));
+                addProductActivity.addToMap(PARAM_ORIGIN, getValues(originOfIngredients));
             }
             if (!manufacturingPlace.getText().toString().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_MANUFACTURING_PLACE, manufacturingPlace.getText().toString());
+                addProductActivity.addToMap(PARAM_MANUFACTURING_PLACE, manufacturingPlace.getText().toString());
             }
             if (!embCode.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_EMB_CODE, getValues(embCode));
+                addProductActivity.addToMap(PARAM_EMB_CODE, getValues(embCode));
             }
             if (!link.getText().toString().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_LINK, link.getText().toString());
+                addProductActivity.addToMap(PARAM_LINK, link.getText().toString());
             }
             if (!countryWherePurchased.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_PURCHASE, getValues(countryWherePurchased));
+                addProductActivity.addToMap(PARAM_PURCHASE, getValues(countryWherePurchased));
             }
             if (!stores.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_STORE, getValues(stores));
+                addProductActivity.addToMap(PARAM_STORE, getValues(stores));
             }
             if (!countriesWhereSold.getChipValues().isEmpty()) {
-                ((AddProductActivity) activity).addToMap(PARAM_COUNTRIES, getValues(countriesWhereSold));
+                addProductActivity.addToMap(PARAM_COUNTRIES, getValues(countriesWhereSold));
             }
         }
     }
@@ -792,7 +773,7 @@ public class AddProductOverviewFragment extends BaseFragment {
      * Chipifies all existing plain text in all the NachoTextViews.
      */
     private void chipifyAllUnterminatedTokens() {
-        NachoTextView nachoTextViews[] = {brand, packaging, categories, label, originOfIngredients, embCode, countryWherePurchased, stores, countriesWhereSold};
+        NachoTextView[] nachoTextViews = {brand, packaging, categories, label, originOfIngredients, embCode, countryWherePurchased, stores, countriesWhereSold};
         for (NachoTextView nachoTextView : nachoTextViews) {
             nachoTextView.chipifyAllUnterminatedTokens();
         }

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/BaseFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/BaseFragment.java
@@ -10,7 +10,6 @@ import android.support.v4.widget.SwipeRefreshLayout;
 import android.view.LayoutInflater;
 import android.view.View;
 import android.view.ViewGroup;
-
 import butterknife.ButterKnife;
 import openfoodfacts.github.scrachx.openfood.R;
 import openfoodfacts.github.scrachx.openfood.models.State;
@@ -64,10 +63,10 @@ public abstract class BaseFragment extends Fragment implements SwipeRefreshLayou
         }
     }
 
-    protected State getStateFromActivityIntent(){
+    protected State getStateFromActivityIntent() {
         Intent intent = getActivity().getIntent();
-        if(intent!=null && intent.getExtras()!=null && intent.getExtras().getSerializable("state")!=null) {
-            return  (State) intent.getExtras().getSerializable("state");
+        if (intent != null && intent.getExtras() != null && intent.getExtras().getSerializable("state") != null) {
+            return (State) intent.getExtras().getSerializable("state");
         }
         return ProductFragment.mState;
     }

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/BaseFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/BaseFragment.java
@@ -1,6 +1,7 @@
 package openfoodfacts.github.scrachx.openfood.fragments;
 
 import android.content.Context;
+import android.content.Intent;
 import android.os.Bundle;
 import android.support.annotation.NonNull;
 import android.support.annotation.Nullable;
@@ -15,6 +16,7 @@ import openfoodfacts.github.scrachx.openfood.R;
 import openfoodfacts.github.scrachx.openfood.models.State;
 import openfoodfacts.github.scrachx.openfood.views.listeners.OnRefreshListener;
 import openfoodfacts.github.scrachx.openfood.views.listeners.OnRefreshView;
+import openfoodfacts.github.scrachx.openfood.views.product.ProductFragment;
 
 public abstract class BaseFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener, OnRefreshView {
 
@@ -60,5 +62,13 @@ public abstract class BaseFragment extends Fragment implements SwipeRefreshLayou
         if (swipeRefreshLayout != null) {
             swipeRefreshLayout.setRefreshing(false);
         }
+    }
+
+    protected State getStateFromActivityIntent(){
+        Intent intent = getActivity().getIntent();
+        if(intent!=null && intent.getExtras()!=null && intent.getExtras().getSerializable("state")!=null) {
+            return  (State) intent.getExtras().getSerializable("state");
+        }
+        return ProductFragment.mState;
     }
 }

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/ContributorsFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/ContributorsFragment.java
@@ -50,12 +50,7 @@ public class ContributorsFragment extends BaseFragment {
     @Override
     public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
-        Intent intent = getActivity().getIntent();
-        if(intent!=null && intent.getExtras()!=null && intent.getExtras().getSerializable("state")!=null) {
-            mState = (State) intent.getExtras().getSerializable("state");
-        }else{
-            mState = ProductFragment.mState;
-        }
+        mState=getStateFromActivityIntent();
         refreshView(mState);
     }
 
@@ -150,118 +145,6 @@ public class ContributorsFragment extends BaseFragment {
         return spannableStringBuilder;
     }
 
-
-    /**  @Override public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
-    super.onViewCreated(view, savedInstanceState);
-    Intent intent = getActivity().getIntent();
-    mState = (State) intent.getExtras().getSerializable("state");
-    final Product product = mState.getProduct();
-
-    if (isNotBlank(product.getCreator())) {
-
-    String[] createdDate = getDateTime(product.getCreatedDateTime());
-    String creatorTxt = getString(R.string.creator_history, createdDate[0], createdDate[1]);
-    creatorText.setMovementMethod(LinkMovementMethod.getInstance());
-    creatorText.setText(creatorTxt + " ");
-    creatorText.append(getContributorsTag(product.getCreator()));
-
-    } else {
-    creatorText.setVisibility(View.INVISIBLE);
-    }
-
-
-    if (isNotBlank(product.getLastModifiedBy())) {
-    String[] lastEditDate = getDateTime(product.getLastModifiedTime());
-    String editorTxt = getString(R.string.last_editor_history, lastEditDate[0], lastEditDate[1]);
-    lastEditorText.setMovementMethod(LinkMovementMethod.getInstance());
-    lastEditorText.setText(editorTxt + " ");
-    lastEditorText.append(getContributorsTag(product.getLastModifiedBy()));
-
-    } else {
-    lastEditorText.setVisibility(View.INVISIBLE);
-    }
-
-
-    if (product.getStatesTags().size() != 0) {
-    String otherEditorsTxt = getString(R.string.other_editors);
-    otherEditorsText.setMovementMethod(LinkMovementMethod.getInstance());
-    otherEditorsText.setText(otherEditorsTxt + " ");
-
-    for (int i = 0; i < product.getEditors().size() - 1; i++) {
-    otherEditorsText.append(getContributorsTag(product.getEditors().get(i)));
-    otherEditorsText.append(",");
-    }
-    otherEditorsText.append(getContributorsTag(product.getEditors().get(product.getEditors().size() - 1)));
-    } else {
-    otherEditorsText.setVisibility(View.INVISIBLE);
-    }
-
-
-    if (!product.getStatesTags().equals("")) {
-
-    statesText.setMovementMethod(LinkMovementMethod.getInstance());
-    statesText.setText("");
-    for (int i = 0; i < product.getStatesTags().size(); i++) {
-    statesText.append(getStatesTag(product.getStatesTags().get(i).split(":")[1]));
-    statesText.append("\n ");
-    }
-    }
-
-    if (isNotBlank(product.getImageFrontUrl())) {
-    Picasso.with(getContext()).load(product.getImageFrontUrl()).into(imgFront);
-    }
-
-    if (isNotBlank(product.getImageIngredientsUrl())) {
-    Picasso.with(getContext()).load(product.getImageIngredientsUrl()).into(imgIngredients);
-    }
-
-    if (isNotBlank(product.getImageNutritionUrl())) {
-    Picasso.with(getContext()).load(product.getImageNutritionUrl()).into(imgNutrients);
-    }
-
-    }
-
-    private String[] getDateTime(String dateTime) {
-    long unixSeconds = Long.valueOf(dateTime);
-    Date date = new java.util.Date(unixSeconds * 1000L);
-    SimpleDateFormat sdf = new java.text.SimpleDateFormat("MMMM dd, yyyy");
-    SimpleDateFormat sdf2 = new java.text.SimpleDateFormat("HH:mm:ss a");
-    sdf2.setTimeZone(java.util.TimeZone.getTimeZone("CET"));
-    String[] formattedDates = new String[]{sdf.format(date), sdf2.format(date)};
-    return formattedDates;
-    }
-
-
-    private CharSequence getContributorsTag(String contributor) {
-    SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder();
-    ClickableSpan clickableSpan = new ClickableSpan() {
-    @Override public void onClick(View view) {
-
-    ProductBrowsingListActivity.startActivity(getContext(), contributor, SearchType.CONTRIBUTOR);
-
-    }
-    };
-    spannableStringBuilder.append(contributor);
-    spannableStringBuilder.setSpan(clickableSpan, 0, spannableStringBuilder.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
-    spannableStringBuilder.append(" ");
-    return spannableStringBuilder;
-    }
-
-
-    private CharSequence getStatesTag(String state) {
-    SpannableStringBuilder spannableStringBuilder = new SpannableStringBuilder();
-    ClickableSpan clickableSpan = new ClickableSpan() {
-    @Override public void onClick(View view) {
-
-    ProductBrowsingListActivity.startActivity(getContext(), state, SearchType.STATE);
-
-    }
-    };
-    spannableStringBuilder.append(state);
-    spannableStringBuilder.setSpan(clickableSpan, 0, spannableStringBuilder.length(), SPAN_EXCLUSIVE_EXCLUSIVE);
-    spannableStringBuilder.append(" ");
-    return spannableStringBuilder;
-    }**/
 
 }
 

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/ProductPhotosFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/fragments/ProductPhotosFragment.java
@@ -56,13 +56,7 @@ public class ProductPhotosFragment extends BaseFragment implements ImagesAdapter
     @Override
     public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
-        Intent intent = getActivity().getIntent();
-        State state;
-        if (intent != null && intent.getExtras() != null && intent.getExtras().getSerializable("state") != null) {
-            state = (State) intent.getExtras().getSerializable("state");
-        } else {
-            state = ProductFragment.mState;
-        }
+        State state=getStateFromActivityIntent();
         product = state.getProduct();
         mFragment = this;
         // initialize the arraylist

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/adapters/EmbCodeAutoCompleteAdapter.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/adapters/EmbCodeAutoCompleteAdapter.java
@@ -13,6 +13,7 @@ import io.reactivex.SingleObserver;
 import io.reactivex.disposables.Disposable;
 import openfoodfacts.github.scrachx.openfood.network.CommonApiManager;
 import openfoodfacts.github.scrachx.openfood.network.OpenFoodAPIService;
+import org.apache.commons.lang3.StringUtils;
 
 public class EmbCodeAutoCompleteAdapter extends ArrayAdapter<String> implements Filterable {
 
@@ -31,6 +32,9 @@ public class EmbCodeAutoCompleteAdapter extends ArrayAdapter<String> implements 
 
     @Override
     public String getItem(int position) {
+        if(position<0|| position>=mEMBCodeList.size()){
+            return StringUtils.EMPTY;
+        }
         return mEMBCodeList.get(position);
     }
 

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/environment/EnvironmentProductFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/environment/EnvironmentProductFragment.java
@@ -49,11 +49,7 @@ public class EnvironmentProductFragment extends BaseFragment {
     @Override
     public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
-        Intent intent = getActivity().getIntent();
-        if(intent!=null && intent.getExtras()!=null && intent.getExtras().getSerializable("state")!=null)
-            mState = (State) intent.getExtras().getSerializable("state");
-        else
-            mState = ProductFragment.mState;
+        mState=getStateFromActivityIntent();
 
         final Product product = mState.getProduct();
         Nutriments nutriments = product.getNutriments();

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/ingredients/IngredientsProductFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/ingredients/IngredientsProductFragment.java
@@ -20,7 +20,6 @@ import android.support.v4.content.ContextCompat;
 import android.support.v4.view.ViewPager;
 import android.support.v7.preference.PreferenceManager;
 import android.support.v7.widget.CardView;
-import android.text.SpannableString;
 import android.text.SpannableStringBuilder;
 import android.text.TextUtils;
 import android.text.method.LinkMovementMethod;
@@ -74,7 +73,7 @@ import openfoodfacts.github.scrachx.openfood.views.adapters.ProductFragmentPager
 import openfoodfacts.github.scrachx.openfood.views.customtabs.CustomTabActivityHelper;
 import openfoodfacts.github.scrachx.openfood.views.customtabs.CustomTabsHelper;
 import openfoodfacts.github.scrachx.openfood.views.customtabs.WebViewFallback;
-import openfoodfacts.github.scrachx.openfood.views.product.ProductFragment;
+import org.apache.commons.lang3.StringUtils;
 import pl.aprilapps.easyphotopicker.DefaultCallback;
 import pl.aprilapps.easyphotopicker.EasyImage;
 
@@ -177,11 +176,7 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
         customTabActivityHelper = new CustomTabActivityHelper();
         customTabsIntent = CustomTabsHelper.getCustomTabsIntent(getContext(), customTabActivityHelper.getSession());
 
-        Intent intent = getActivity().getIntent();
-        if(intent!=null && intent.getExtras()!=null && intent.getExtras().getSerializable("state")!=null)
-            mState = (State) intent.getExtras().getSerializable("state");
-        else
-            mState = ProductFragment.mState;
+        mState=getStateFromActivityIntent();
         product = mState.getProduct();
 
         presenter = new IngredientsProductPresenter(product, this);
@@ -200,11 +195,7 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
     @Override
     public void onViewCreated(View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
-        Intent intent = getActivity().getIntent();
-        if(intent!=null && intent.getExtras()!=null && intent.getExtras().getSerializable("state")!=null)
-            mState = (State) intent.getExtras().getSerializable("state");
-        else
-            mState = ProductFragment.mState;
+        mState=getStateFromActivityIntent();
         refreshView(mState);
     }
 
@@ -233,57 +224,29 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
         List<String> aminoAcidTagsList = product.getAminoAcidTags();
         List<String> mineralTags = product.getMineralTags();
         List<String> otherNutritionTags = product.getOtherNutritionTags();
-        String prefix = " ";
 
         if (!vitaminTagsList.isEmpty()) {
-            StringBuilder vitaminStringBuilder = new StringBuilder();
             vitaminsTagsTextCardView.setVisibility(View.VISIBLE);
             vitaminTagsTextView.setText(bold(getString(R.string.vitamin_tags_text)));
-            for (String vitamins : vitaminTagsList) {
-                vitaminStringBuilder.append(prefix);
-                prefix = ", ";
-                vitaminStringBuilder.append(trimLanguagePartFromString(vitamins));
-            }
-            vitaminTagsTextView.append(vitaminStringBuilder.toString());
+            vitaminTagsTextView.append(buildStringBuilder(vitaminTagsList, StringUtils.SPACE));
         }
 
         if (!aminoAcidTagsList.isEmpty()) {
-            String aminoPrefix = " ";
-            StringBuilder aminoAcidStringBuilder = new StringBuilder();
             aminoAcidTagsTextCardView.setVisibility(View.VISIBLE);
             aminoAcidTagsTextView.setText(bold(getString(R.string.amino_acid_tags_text)));
-            for (String aminoAcid : aminoAcidTagsList) {
-                aminoAcidStringBuilder.append(aminoPrefix);
-                aminoPrefix = ", ";
-                aminoAcidStringBuilder.append(trimLanguagePartFromString(aminoAcid));
-            }
-            aminoAcidTagsTextView.append(aminoAcidStringBuilder.toString());
+            aminoAcidTagsTextView.append(buildStringBuilder(aminoAcidTagsList, StringUtils.SPACE));
         }
 
         if (!mineralTags.isEmpty()) {
-            String mineralPrefix = " ";
-            StringBuilder mineralsStringBuilder = new StringBuilder();
             mineralTagsTextCardView.setVisibility(View.VISIBLE);
             mineralTagsTextView.setText(bold(getString(R.string.mineral_tags_text)));
-            for (String mineral : mineralTags) {
-                mineralsStringBuilder.append(mineralPrefix);
-                mineralPrefix = ", ";
-                mineralsStringBuilder.append(trimLanguagePartFromString(mineral));
-            }
-            mineralTagsTextView.append(mineralsStringBuilder);
+            mineralTagsTextView.append(buildStringBuilder(mineralTags, StringUtils.SPACE));
         }
 
         if (!otherNutritionTags.isEmpty()) {
-            String otherNutritionPrefix = " ";
-            StringBuilder otherNutritionStringBuilder = new StringBuilder();
             otherNutritionTagTextView.setVisibility(View.VISIBLE);
             otherNutritionTagTextView.setText(bold(getString(R.string.other_tags_text)));
-            for (String otherSubstance : otherNutritionTags) {
-                otherNutritionStringBuilder.append(otherNutritionPrefix);
-                otherNutritionPrefix = ", ";
-                otherNutritionStringBuilder.append(trimLanguagePartFromString(otherSubstance));
-            }
-            otherNutritionTagTextView.append(otherNutritionStringBuilder.toString());
+            otherNutritionTagTextView.append(buildStringBuilder(otherNutritionTags, StringUtils.SPACE));
         }
 
         additiveProduct.setText(bold(getString(R.string.txtAdditives)));
@@ -339,7 +302,7 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
             traceProduct.append(" ");
 
             String trace;
-            String traces[] = product.getTraces().split(",");
+            String[] traces = product.getTraces().split(",");
             for (int i = 0; i < traces.length - 1; i++) {
                 trace = traces[i];
                 traceProduct.append(Utils.getClickableText(trace, trace, SearchType.TRACE, getActivity(), customTabsIntent));
@@ -383,6 +346,16 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
         } else {
             novaLayout.setVisibility(View.GONE);
         }
+    }
+
+    private StringBuilder buildStringBuilder(List<String> stringList, String prefix) {
+        StringBuilder otherNutritionStringBuilder = new StringBuilder();
+        for (String otherSubstance : stringList) {
+            otherNutritionStringBuilder.append(prefix);
+            prefix = ", ";
+            otherNutritionStringBuilder.append(trimLanguagePartFromString(otherSubstance));
+        }
+        return otherNutritionStringBuilder;
     }
 
     private CharSequence getAdditiveTag(AdditiveName additive) {
@@ -532,9 +505,6 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
         return ssb;
     }
 
-    private SpannableString buildAdditivesList(List<AdditiveName> additives) {
-        return null;
-    }
 
     @Override
     public void showAdditives(List<AdditiveName> additives) {
@@ -610,7 +580,7 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
             }
             else
             {
-                mState = (State) getActivity().getIntent().getExtras().getSerializable( "state" );
+                mState = getStateFromActivityIntent();
                 Intent intent = new Intent( getContext(), AddProductActivity.class );
                 intent.putExtra("send_updated", sendUpdatedIngredientsImage);
                 intent.putExtra( "edit_product", mState.getProduct() );
@@ -714,7 +684,6 @@ public class IngredientsProductFragment extends BaseFragment implements IIngredi
                 ActivityCompat.requestPermissions(getActivity(), new String[]{CAMERA}, MY_PERMISSIONS_REQUEST_CAMERA);
             } else {
                 EasyImage.openCamera(this, 0);
-//                EasyImage.openGallery(this);
             }
         }
     }

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/nutrition/NutritionProductFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/nutrition/NutritionProductFragment.java
@@ -128,15 +128,10 @@ public class NutritionProductFragment extends BaseFragment implements CustomTabA
     @Override
     public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
-        Intent intent = getActivity().getIntent();
         // use VERTICAL divider
         DividerItemDecoration dividerItemDecoration = new DividerItemDecoration(nutrimentsRecyclerView.getContext(), VERTICAL);
         nutrimentsRecyclerView.addItemDecoration(dividerItemDecoration);
-        if (intent != null && intent.getExtras() != null && intent.getExtras().getSerializable("state") != null) {
-            refreshView((State) intent.getExtras().getSerializable("state"));
-        } else {
-            refreshView(ProductFragment.mState);
-        }
+        refreshView(getStateFromActivityIntent());
     }
 
     @Override

--- a/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/SummaryProductFragment.java
+++ b/app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/SummaryProductFragment.java
@@ -171,12 +171,7 @@ public class SummaryProductFragment extends BaseFragment implements CustomTabAct
         customTabActivityHelper.setConnectionCallback(this);
         customTabsIntent = CustomTabsHelper.getCustomTabsIntent(getContext(), customTabActivityHelper.getSession());
 
-        Intent intent = getActivity().getIntent();
-        if (intent.getExtras() != null && intent.getExtras().getSerializable("state") != null) {
-            state = (State) intent.getExtras().getSerializable("state");
-        } else {
-            state = ProductFragment.mState;
-        }
+        state=getStateFromActivityIntent();
 
         presenter = new SummaryProductPresenter(product, this);
     }

